FIX: fix interpolate with kwarg limit area and limit direction using pad or bfill

cchwala opened the pull request:
- [x] closes #26796
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This separates a large part of the code changes from #25141 to keep things comprehendible.

cchwala committed:
Added failing test for https://github.com/pandas-dev/pandas/issues/26796

This test currently only test `limit_area`.

For `limit_direction` the implementation should later raise an error,
because `pad` and `bfill` both already define a direction. But let's
now first do the implementation of the `limit_area` for `pad`
and `bfill`.

cchwala committed:
Added implementation to support `limit_area`

Since methods `pad` and `bfill` in `blocks.interpolate` end up
using `missing.interpolate_2d` which can not (easily) be extended
to support `limit_area`, I introduce the new function
`missing.interpolate_1d_fill`. It is a modified copy of `interpolate_2d`
but only works for 1d data and uses newly introduced function
`_derive_indices_of_nans_to_preserve`, which is now also used in
`missing.interpolate_1d`. It works the same way as the
1D-interpolation functions which are based on scipy-interpolation which
are applied via np.apply_along_axis.

cchwala committed:
fix test

pep8speaks commented:
Hello @cchwala! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/core/missing.py`](https://github.com/pandas-dev/pandas/blob/fd5d8e845827b185acd387b262041f7ec9d2434e/pandas/core/missing.py):

> [Line 290:1](https://github.com/pandas-dev/pandas/blob/fd5d8e845827b185acd387b262041f7ec9d2434e/pandas/core/missing.py#L290): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 499:1](https://github.com/pandas-dev/pandas/blob/fd5d8e845827b185acd387b262041f7ec9d2434e/pandas/core/missing.py#L499): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

WillAyd commented:
@cchwala so IIUC you are doing a precursor to solve the issue with pad and `limit_area` and then planning to add `max_gap` after that gets merged in the original PR right?

cchwala commented:
@WillAyd Yes. That is my plan. 

I am reusing the code from the PR with `max_gap` that actually solves #26796. The `max_gap` PR will be a lot smaller when the changes from this PR are merge.

cchwala committed:
pep8

cchwala committed:
fixed small error that actually had no effect since the input array `values` also was changed via appliying `func`

cchwala committed:
Raise when forbidden combination of `method` and `limit_direction` are used

Test for all forbidden combos of `pad` and `backfill` is included

cchwala committed:
Updated docstring with info about allowed combinations of `method` and `limit_direction`

cchwala committed:
clean up

* black formatting
* typo

WillAyd commented on pandas/core/missing.py:
```diff
@@ -477,6 +496,65 @@ def _akima_interpolate(xi, yi, x, der=0, axis=0):
 
+def interpolate_1d_fill(
+    values,
+    method="pad",
```
```suggestion
    method: str = "pad",
```

WillAyd commented on pandas/core/missing.py:
```diff
@@ -477,6 +496,65 @@ def _akima_interpolate(xi, yi, x, der=0, axis=0):
+def interpolate_1d_fill(
+    values,
+    method="pad",
+    axis=0,
```
```suggestion
    axis: Axis = 0,
```

(import from pandas._typing)

WillAyd commented on pandas/core/missing.py:
```diff
@@ -477,6 +496,65 @@ def _akima_interpolate(xi, yi, x, der=0, axis=0):
+    values,
+    method="pad",
+    axis=0,
+    limit=None,
```
```suggestion
    limit: Optional[int] = None,
```

WillAyd commented on pandas/core/missing.py:
```diff
@@ -477,6 +496,65 @@ def _akima_interpolate(xi, yi, x, der=0, axis=0):
+    method="pad",
+    axis=0,
+    limit=None,
+    limit_area=None,
```
```suggestion
    limit_area: Optional[str] = None,
```

WillAyd commented on pandas/core/missing.py:
```diff
@@ -477,6 +496,65 @@ def _akima_interpolate(xi, yi, x, der=0, axis=0):
+    limit=None,
+    limit_area=None,
+    fill_value=None,
+    dtype=None,
```
```suggestion
    dtype: Optional[Dtype] = None,
```

(import from pandas._typing)

WillAyd commented on pandas/core/missing.py:
```diff
@@ -477,6 +496,65 @@ def _akima_interpolate(xi, yi, x, der=0, axis=0):
+    axis=0,
+    limit=None,
+    limit_area=None,
+    fill_value=None,
```
```suggestion
    fill_value: Optional[Hashable] = None,
```

WillAyd reviewed (changes requested):
Looks pretty good. If you can annotate any new functions would be helpful. Added what I think should be

cchwala committed:
Added entry to whatsnew file

cchwala committed:
Removed `axis` kwarg from `interpolate_1d_fill` because it was unused

cchwala committed:
Type annotations added to new function `interpolate_1d_fill`

cchwala committed:
fixed incorrectly sorted imports

cchwala commented:
Should be ready now. I am offline for some hours of sleep, though.

cchwala commented:
@WillAyd Any further request for changes?

WillAyd commented on pandas/core/generic.py:
```diff
@@ -6527,9 +6527,11 @@ def replace(
         inplace : bool, default False
             Update the data in place if possible.
-        limit_direction : {'forward', 'backward', 'both'}, default 'forward'
+        limit_direction : {'forward', 'backward', 'both'}, default is None
```
Should revert this; though it shows `None` as the value it still does default to `forward` no?

cchwala replied:
Thanks for spotting this. Will be fixed in next series of commits.

jreback commented on pandas/core/missing.py:
```diff
@@ -314,6 +289,53 @@ def interpolate_1d(
 
 
+def _derive_indices_of_nans_to_preserve(
+    yvalues, valid, invalid, limit, limit_area, limit_direction
```
can you try to type these * add a Paramaters / Returns section

cchwala replied:
Yes. Will be update in next series of commits

jreback commented on pandas/core/internals/blocks.py:
```diff
@@ -1168,16 +1170,36 @@ def _interpolate_with_fill(
+        # 2. All other cases: Then, `missing.interpolate_2d()` can be used.
+        if limit_area is not None:
+
+            def func(x):
```
instead of handling this case here and distributing the logic, can you simply call it from interpolate_2d?

cchwala replied:
Yes. Will be moved in next series of commits.

jreback commented on pandas/core/generic.py:
```diff
@@ -6720,6 +6722,28 @@ def interpolate(
                 "column to a numeric dtype."
             )
 
+        # Set `limit_direction` depending on `method`
```
is there a reason this logic does not simply belong in interpolate_2d?

jreback replied:
or call something like clean_fill_method

cchwala replied:
`limit_direction` has no effect in `interpolate_2d`.

My idea was to detect conflicting user input, e.g. `pad` with `limit_direction='backward'` as early as possible in the chain of calls to the main functions that are involved.

But I can move the logic to a function like `clean_fill_method`, e.g. called `set_limit_direction`. Should this function belong to generic.py or missing.py?

simonjayhawkins replied:
interpolate_2d is also called from fillna. only NDFrame.interpolate allows limit_direction so does make sense to validate here.

clean_fill_method checks method. probably want to avoid passing more parameters around.

jreback replied:
pls make a method similr to clean_fill_method then, e.g this logic should living pandas/core/missing.py

cchwala replied:
not sure what to do here. after merge with master, there is similar code at this location, but it is not mine...

jreback commented on pandas/core/missing.py:
```diff
@@ -478,6 +500,66 @@ def _akima_interpolate(xi, yi, x, der=0, axis=0):
         return [P(x, nu) for nu in der]
 
 
+def interpolate_1d_fill(
```
can you move this under interpolate_1d

cchwala replied:
`interpolate_1d_fill` is a special case needed by `blocks._interpolate_with_fill()` to support the limit kwargs with method `pad`. As you requested in the comment above `interpolate_1d_fill` is now called from within `interpolate_2d` which is what `blocks._interpolate_with_fill()` calls.

You request, moving it to `missing.interpolate_1d` will, in my opinion, not work, because `missing.interpolate_1d` is not reached in the call sequence for interpolating with method `pad` since `blocks.Block.interpolate()` splits up into `blocks.Block._interpolate` (for scipy wrappers) and `blocks.Block._interpolate_with_fill` (for pad methods). `missing.interpolate_1d` is only called in the former case, i.e. for methods using the scipy wrappers.

simonjayhawkins replied:
@cchwala interpolate_2d will work on a 1d array. did you investigate applying it along an axis (with masking logic) rather than creating a 1d version.

simonjayhawkins replied:
see #34749 for alternative implementation calling interpolate_2d instead. interpolate_2d already has the limit logic so no need to use the preserve_nans set based logic.

cchwala replied:
I used the nice solution from #34749 as suggested above

cchwala replied:
as a consequence there is no `interpolate_1d_fill` anymore

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -1040,7 +1040,7 @@ Indexing
 ^^^^^^^
 
--
+-  Bug in :meth:`Series.interpolate` where kwarg ``limit_area`` and ``limit_direction`` had no effect when using methods ``pad`` and `backfill`` (:issue:`31048`)
```
move this to 1.1

cchwala replied:
done

jreback reviewed (changes requested):
looks pretty good

cchwala committed:
Added type annotation, updated docstring and removed unnecessary arguments

Test on my local machine are not affected by removing the unncessery
arguments `valid` and `invalid`, which are now derived within the
function.

WillAyd commented:
@cchwala can you address comments?

cchwala committed:
Reverting docstring entry for default value of `limit_direction`

cchwala committed:
Moved logic for calling `missing.interpolate_1d_fill` to `missing.interpolate_2d`

cchwala committed:
Moved whatsnew entry to v1.1.0.rst

cchwala committed:
clean up

-black formating
-remove variables that are now obsolete

cchwala commented:
@WillAyd Sorry for the delay. I adjusted the code according to most comments. I am now looking into the last two open comments and will respond/commit when I have an answer.

cchwala committed:
fixed missing Optional in type definition

cchwala commented on pandas/core/missing.py:
```diff
@@ -314,6 +287,73 @@ def interpolate_1d(
+        preserve_nans |= mid_nans
+
+    # sort preserve_nans and covert to list
+    preserve_nans = sorted(preserve_nans)
```
@jreback Should I force a set here with
```
preserve_nans = set(sorted(preserve_nans))
```
so that the CI type validation does not complain?

WillAyd replied:
If you don't re-use the same variable should be OK, so something like `nans_list = sorted(preserve_nans)` should work

cchwala replied:
Solved

cchwala commented:
@WillAyd @jreback Can you comment on the changes. I would potentially have time to further work on this and the related https://github.com/pandas-dev/pandas/pull/25141, which has to wait till this is merged.

cchwala committed:
small fix so that CI type validation does not complain

cchwala committed:
Merge remote-tracking branch 'upstream/master' into fix_interpolate_limit_area_and_limit_direction_with_pad

Manually resolved conflicts:
 * doc/source/whatsnew/v1.1.0.rst: Just a conflicting line
 * pandas/tests/series/test_missing.py: All test for interpolation have
   been moved to a new file pandas/tests/series/methods/test_interpolate.py
   I accepted all changes from upstream, which removed my tests from
   test_missing.py and manually added my test to the new file
   methods/test_interpolate.py

cchwala commented:
@WillAyd I resolved the conflicts with upstream/master so that this PR can potentially be merged.

WillAyd commented on pandas/core/missing.py:
```diff
@@ -314,6 +287,73 @@ def interpolate_1d(
+
+    Returns
+    -------
+    preserve_nans: set
```
```suggestion
    preserve_nans: list of int
```

sorted converts the set to a list

WillAyd commented on pandas/core/missing.py:
```diff
@@ -314,6 +287,73 @@ def interpolate_1d(
+    limit: Optional[int] = None,
+    limit_area: Optional[str] = None,
+    limit_direction: Optional[str] = None,
+):
```
```suggestion
) -> List[int]:
```

cchwala committed:
Apply suggestions from code review concerning list instead of set

Co-Authored-By: William Ayd <william.ayd@icloud.com>

cchwala committed:
added import for missing List type

cchwala committed:
fixed unsorted order of imports

cchwala commented:
@WillAyd Sorry for missing the details of the required changes. These are solved now.

WillAyd commented on pandas/tests/series/methods/test_interpolate.py:
```diff
@@ -400,6 +400,75 @@ def test_interp_limit_area(self):
+
+    def test_interp_raise_limit_direction_and_pad_or_bfill(self):
+        s = Series([1, 2, 3])
+        forbidden_combinations = [
```
These should really be parametrized...but not going to block since this PR has gone on for a long time (great job on keeping updated by the way)

simonjayhawkins replied:
done in #34746

WillAyd reviewed (approved):
small comment that can be address in a follow on but otherwise lgtm. @jreback

jreback commented:
@cchwala can you merge master.

simonjayhawkins commented:
@cchwala can you merge master

simonjayhawkins commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -308,7 +308,7 @@ Missing
 
 - Calling :meth:`fillna` on an empty Series now correctly returns a shallow copied object. The behaviour is now consistent with :class:`Index`, :class:`DataFrame` and a non-empty :class:`Series` (:issue:`32543`).
-
+- Bug in :meth:`Series.interpolate` where kwarg ``limit_area`` and ``limit_direction`` had no effect when using methods ``pad`` and `backfill`` (:issue:`31048`)
```
issue #26796

jreback replied:
pls move to 1.2

simonjayhawkins commented on pandas/core/generic.py:
```diff
@@ -6671,9 +6671,11 @@ def replace(
         inplace : bool, default False
             Update the data in place if possible.
-        limit_direction : {'forward', 'backward', 'both'}, default 'forward'
+        limit_direction : {'forward', 'backward', 'both'}, default is 'forward'
```
the default is now 'backward' for method 'backfill' and 'bfill'?

jreback replied:
@cchwala can you address

cchwala replied:
already change after merging master

jreback commented:
@simonjayhawkins I think this is worth merge *after* your set cleaning in interpolate.

simonjayhawkins commented:
> @simonjayhawkins I think this is worth merge _after_ your set cleaning in interpolate.

i've done this locally but get
```
$ git push
fatal: The upstream branch of your current branch does not match
the name of your current branch.  To push to the upstream branch
on the remote, use

    git push upstream HEAD:refs/pull/31048/head

To push to the branch of the same name on the remote, use

    git push upstream fix_interpolate_limit_area_and_limit_direction_with_pad

To choose either option permanently, see push.default in 'git help config'.
```

am using hub.

BUT tests are failing. I think related to changes in #33959. not looked yet.

simonjayhawkins commented:
> BUT tests are failing. I think related to changes in #33959. not looked yet.

will need to make some changes to the implementation of #33959, will raise a PR shortly based on changes in #34749

cchwala commented:
@simonjayhawkins Sorry for not responding. I was, and still am, very busy with other things. I am available to help with this PR from July on.

jreback commented on pandas/core/generic.py:
```diff
@@ -6824,7 +6826,7 @@ def interpolate(
         limit=None,
         inplace=False,
-        limit_direction="forward",
+        limit_direction=None,
```
can you type this (other keywords are a +)

cchwala replied:
not relevant anymore

jreback commented on pandas/core/missing.py:
```diff
@@ -314,6 +287,73 @@ def interpolate_1d(
+    valid = ~invalid
+
+    # These are sets of index pointers to invalid values... i.e. {0, 1, etc...
+    all_nans = set(np.flatnonzero(invalid))
```
better to use numpy set ops here, no?

e.g. np.unique

cchwala replied:
not relevant anymore

jreback reviewed (changes requested):
not really sure of the status here and @simonjayhawkins had some cleanups as well, can we clarify.

simonjayhawkins commented:
> not really sure of the status here and @simonjayhawkins had some cleanups as well, can we clarify.

I did break some bits off this PR. so does need a merge of master to see what's left (I did try this https://github.com/pandas-dev/pandas/pull/31048#issuecomment-643394779)

cchwala commented:
Okay. I will do the merge with master today to see what is left. I guess, at least the tests will still be of value.

cchwala committed:
Merge remote-tracking branch 'upstream/master' into fix_interpolate_limit_area_and_limit_direction_with_pad

I removed most of the code that I have added in this branch and will add it back in step by step in the next commits

cchwala committed:
Added tests back in

Tests should be red now

cchwala commented:
I am now looking into [the suggestion](https://github.com/pandas-dev/pandas/pull/31048/files/e259549e2f369bd07a731681d1278e1e5d605007#r439754395) from @simonjayhawkins to use `interpolate_2d` instead of introducing `interpolate_1d_fill`

cchwala committed:
Added new solution to account for limit_area with pad

provided by https://github.com/pandas-dev/pandas/pull/34749

Test are green now

cchwala committed:
black formatting

cchwala committed:
added whatsnew entry

cchwala commented:
@jreback can you please have another look? As expected there is not much left from this PR because parts of it have already been merge elsewhere. Most of your code review comments do not apply anymore, only [this one](https://github.com/pandas-dev/pandas/pull/31048/files/e259549e2f369bd07a731681d1278e1e5d605007#r477598954).

cchwala commented:
@jreback Sorry to ping you again, but it would be great if I could finish this PR with (hopefully) final adjustments tomorrow (Friday). Otherwise things would have to wait again another two weeks because I will be offline.

jreback commented:
will have to wait as i am away

cchwala commented:
Okay. No problem.

jreback commented on pandas/core/internals/blocks.py:
```diff
@@ -1176,15 +1179,43 @@ def _interpolate_with_fill(
             dtype=self.dtype,
         )
 
+        if limit_area is None:
```
can you put this in a function in `missing` and just call here?

cchwala replied:
Okay. But if I do this I could also directly integrate it into `missing.interpolate_2d`, e.g. like [in this branch](https://github.com/pandas-dev/pandas/compare/master...cchwala:quick_test_interpolate_limit_area). Advantage, no need to introduce yet another function that does some interpolation. Disadvantage, obfuscates `missing.interpolate_2d`. For this, adding some comments would help for sure.

If you prefer the separate function in `missing`, what function name would you choose?

simonjayhawkins replied:
> e.g. like [in this branch](https://github.com/pandas-dev/pandas/compare/master...cchwala:quick_test_interpolate_limit_area).

I would merge recursive solution here for review (so that ci has been run against it, i think needs changes to pandas/core/arrays/categorical.py see https://github.com/pandas-dev/pandas/compare/master...simonjayhawkins:fix_interpolate_limit_area_and_limit_direction_with_pad)

cchwala replied:
Okay. Merged and CI is running.

I have also pushed [a branch](https://github.com/pandas-dev/pandas/compare/master...cchwala:quick_test_limit_area_option2) with an example implementation using a separate function in `missing`. The name of the new function is not good, but the implementation would probably look something like this.

For both solutions, tests for interpolation pass, but I did not run the full test suite. Will see what changes are required when CI is done.

cchwala replied:
Test are green. 

If you find the recursive solution appropriate, I would add a comment to the function that briefly explains the reasoning. Anything else that needs improvement?

simonjayhawkins replied:
@jreback

cchwala committed:
Test with moving logic for interpolate_2d with `limite_area` directly into interpolate_2d

cchwala committed:
fix wrong arg order by using kwargs as suggested in https://github.com/pandas-dev/pandas/commit/75f3804982d912acef078e9bb5ec43806f596954

simonjayhawkins reviewed (changes requested):
Thanks @cchwala can you add the type annotations as requested by @WillAyd https://github.com/pandas-dev/pandas/pull/31048/files#r367123208

cchwala committed:
Added comment to explain recursion and added typing for interpolate_2d

cchwala committed:
improved test code coverage

cchwala committed:
fixed wrong typing

simonjayhawkins commented:
@cchwala can you resolve merge conflicts and can take another look

jreback commented on pandas/core/missing.py:
```diff
@@ -543,12 +544,62 @@ def _cubicspline_interpolate(xi, yi, x, axis=0, bc_type="not-a-knot", extrapolat
+    # and `limit_area` logic along a certain axis.
+    if limit_area is not None:
+
+        def func(values):
```
can you make this a module level function and just call it here

jreback reviewed (changes requested):
looks ok, some comments, can you merge master and we can get this in.

jreback commented:
moving off 1.2 need merge master and update to comments.

arw2019 commented:
Closing in favor of #38106
